components/sync15/src/changeset.rs

```diff
@@ -61,43 +61,54 @@ impl OutgoingChangeset {
 impl IncomingChangeset {
     pub fn fetch(
         client: &Sync15StorageClient,
-        state: &GlobalState,
+        state: &mut CollState,
         collection: String,
         collection_request: &CollectionRequest,
     ) -> Result<IncomingChangeset> {
-        let records = client.get_encrypted_records(collection_request)?;
-        let timestamp = state.last_modified_or_zero(&collection);
+        let (records, timestamp) = match client.get_encrypted_records(collection_request)? {
+            Sync15ClientResponse::Success {
+                record,
+                last_modified,
+                ..
+            } => (record, last_modified),
+            other => return Err(other.create_storage_error().into()),
+        };
+        // xxx - duplication below of `timestamp` smells wrong
+        state.last_modified = timestamp;
         let mut result = IncomingChangeset::new(collection, timestamp);
         result.changes.reserve(records.len());
-        let key = state.key_for_collection(&result.collection)?;
         for record in records {
-            // TODO: if we see a HMAC error, may need to update crypto/keys?
-            let decrypted = record.decrypt(&key)?;
+            // if we see a HMAC error, we've made an explicit decision to
+            // NOT handle it here, but restart the global state machine.
+            // That should cause us to re-read crypto/keys and things should
+            // work (although if for some reason crypto/keys was updated but
+            // not all storage was wiped we are probably screwed.)
```

Comment:
Yeah—I think we'll have to wipe everything from the server at that point. It's extreme, but skipping individual records seems scarier for data integrity, especially in a corner case that really shouldn't (🤞) happen. Definitely not something we need to figure out in this bug, though, but thanks for bringing it up in the comment!

Reply:
Yeah, this might bite us, so I opened #991 and tentatively nominated it as "ship ready"